doc/rst/lib_spi.rst

```diff
@@ -823,10 +939,360 @@ As application runs that reads a value from the SPI connected WiFi chip and prin
+To show the bus functioning the demo application also has a tester
+component connected to an SPI master bus which is connected (in
+simulation) to the the SPI slave. This allows the generation of
+traffic to show the the communication functioning.
+
+The application consists of five tasks:
+
+   - A task that controls the SPI slave ports
+
+   - A task that implements the register file handling calls from the
+     SPI slave component and the application
+
+   - An application task that connects to the register file task
+
+   - A task that controls the SPI master ports used for testing
+
+   - A tester task that outputs commands to the SPI master task
+
+:numref:`spi_slave_example_block` shows the task and communication structure of the application.
+
+.. _spi_slave_example_block:
+
+.. figure:: ../images/spi_slave_example_block_diagram.png
+   :width: 90%
+   :align: center
+
+   Block diagram of SPI slave application example
+
+
+These tasks communicate via the use of xC interfaces. :numref:`spi_slave_example_task` shows
+the task and communication structure of the application.
+
+.. _spi_slave_example_task:
+
+.. uml::
+   :width: 60%
+   :caption: Task diagram of SPI slave example
+
+   @startuml
+   circle tester
+   circle SPI_master
+   circle app
+   circle reg_file
+   circle SPI_slave
+
+   ' Force layout order with hidden links
+   SPI_master -[#transparent]-> SPI_slave
+
+   ' Actual displayed nodes
+   tester -r-> SPI_master : spi_master_if
+   SPI_slave -r-> reg_file : spi_slave_callback_if
+   app -l-> reg_file : reg_if
+   @enduml
+
+
+Declaring ports
+...............
+
+The SPI library connects to external pins via xCORE ports. In
+``main.xc`` these are declared as variables of type ``port`` at the
+start of the file:
+
+.. literalinclude:: ../../examples/app_spi_slave/src/main.xc
+    :language: c
+    :start-at: p_sclk
+    :end-at: clock
+
+.. note::
+    
+    There is also a clock declaration since the slave needs to use an internal clock as well as ports inside the xCORE device.
+
+How the ports (e.g. ``XS1_PORT_1I``) relate to external pins will
+depend on the exact device being used. See the device datasheet for details.
+
+This application also has an SPI master interface on different ports:
+
+.. literalinclude:: ../../examples/app_spi_slave/src/main.xc
+    :language: c
+    :start-at: p_test_sclk
+    :end-at: p_test_mosi
+
+|newpage|
+
+The application main() function
+...............................
+
+Below is the source code for the main function of this application,
+which is taken from the source file ``main.xc``
+
+.. literalinclude:: ../../examples/app_spi_slave/src/main.xc
+    :language: c
+    :start-at: int main
+
+Looking at this in a more detail you can see the following:
+
+  - The par functionality describes running five separate tasks in
+    parallel; three are for the main application and two are for the
+    tester.
+
+  - The ``spi_slave`` task controls the application
```

Comment:
same comments on using () for functions. Somewhat personal preference but I think an accepted standard